File: scripts/services/planning_ui/renderers/viewer/neighbourhood.mjs
// Page 2 — a component's dependency neighbourhood, at a depth chosen here.
//
// Backstage's Catalog Graph convention, adopted unchanged: the depth-limited
// neighbourhood of ONE entity, never the whole graph. The walk below is the
// same breadth-first walk as `planning_ui/views/neighbourhood.py::neighbourhood`
// — that copy is the one the tests hold to a hand-known answer; this one
// runs in the browser so a depth change is instant rather than a re-derive.
// Nearest-first order is what the node cap truncates by.

import React from "../vendor/react@19.0.0/react.mjs";
import { ForceGraph } from "./graph.mjs";
import { Legend, StateBadge, Counts } from "./panel.mjs";

const h = React.createElement;
const { useEffect, useMemo, useState } = React;

// The default depth and the node cap, set from the measured legibility
// figure — see the phase doc's runtime verification: the radial layout was
// read clearly at 25 nodes and became a ring of overlapping labels past
// ~30. The cap is a default the reader raises, not a hard split.
export const DEFAULT_DEPTH = 2;
export const DEFAULT_NODE_CAP = 25;

export function walk(index, rootId, depth) {
  const adjacency = new Map();
  const add = (a, b) => { if (!adjacency.has(a)) adjacency.set(a, new Set()); adjacency.get(a).add(b); };
  for (const e of index.edges) { add(e.source, e.target); add(e.target, e.source); }
  const distance = new Map([[rootId, 0]]);
  const order = [rootId];
  const queue = [rootId];
  while (queue.length) {
    const current = queue.shift();
    if (distance.get(current) >= depth) continue;
    const next = [...(adjacency.get(current) || [])].sort();
    for (const n of next) {
      if (distance.has(n)) continue;
      distance.set(n, distance.get(current) + 1);
      order.push(n);
      queue.push(n);
    }
  }
  const edges = index.edges.filter((e) => distance.has(e.source) && distance.has(e.target));
  return { order, distance, edges };
}

function Via({ via, phases }) {
  return h("ul", null, via.map((v, i) => h("li", { key: i, className: "mono" },
    `${v.source.split(":").slice(1).join(":")} → ${v.target.split(":").slice(1).join(":")}`,
    h(StateBadge, { state: v.state }),
    h("div", { className: "muted" }, `${v.file}:${v.line}`),
  )));
}

function NodePanel({ node, index, onNavigate }) {
  const out = index.edges.filter((e) => e.source === node.id);
  const inc = index.edges.filter((e) => e.target === node.id);
  const link = (id) => h("button", { className: "link", onClick: () => onNavigate(id) }, id.split(":").slice(1).join(":"));
  return h("div", null,
    h("h2", null, node.label),
    h("dl", null,
      h("dt", null, "kind"), h("dd", null, node.kind === "missing" ? "missing — the edge resolves to no node" : node.kind),
      node.file ? [h("dt", { key: "f" }, "read from"), h("dd", { key: "fv", className: "mono" }, `${node.file}:${node.line}`)] : null,
      node.kind === "component" ? [
        h("dt", { key: "d" }, "declares"),
        h("dd", { key: "dv" }, node.declares === "nothing" ? "nothing a parser can read — a worklist entry"
          : node.declares === "prose-only" ? "in prose only — a worklist entry"
          : node.declares === "excepted" ? `excepted (${node.disposition})` : node.declares || "—"),
        h("dt", { key: "s" }, "status line"), h("dd", { key: "sv" }, node.status || "—"),
      ] : null,
    ),
    node.kind === "component" && node.owned_phases.length ? [
      h("h3", { key: "oh" }, `own phases (${node.owned_phases.length})`),
      h("ul", { key: "ol" }, node.owned_phases.map((id) => {
        const ph = index.phases[id];
        return h("li", { key: id, className: "mono" }, ph ? ph.label : id, h("span", { className: "badge" }, ph && ph.status ? ph.status : "no status"));
      }),
      ),
    ] : null,
    h("h3", null, `depends on (${out.length})`),
    out.length ? h("ul", null, out.map((e) => h("li", { key: e.target }, link(e.target), h(StateBadge, { state: e.state }), h(Via, { via: e.via, phases: index.phases })))) : h("p", { className: "muted" }, "nothing across a component boundary"),
    h("h3", null, `depended on by (${inc.length})`),
    inc.length ? h("ul", null, inc.map((e) => h("li", { key: e.source }, link(e.source), h(StateBadge, { state: e.state }), h(Via, { via: e.via, phases: index.phases })))) : h("p", { className: "muted" }, "nothing"),
    node.internal && node.internal.length ? [
      h("h3", { key: "ih" }, `within this component (${node.internal.length})`),
      h("p", { key: "ip", className: "muted" }, "one of its phases depending on another — not drawn, because it crosses no boundary"),
      h(Via, { key: "iv", via: node.internal, phases: index.phases }),
    ] : null,
  );
}

export function NeighbourhoodPage({ views, route, navigate }) {
  const index = views.neighbourhood;
  const components = Object.values(index.nodes).filter((n) => n.kind === "component");
  const rootId = route.id && index.nodes[route.id] ? route.id : (components[0] && components[0].id);
  const depth = Number.isFinite(route.depth) ? route.depth : DEFAULT_DEPTH;
  const cap = Number.isFinite(route.cap) ? route.cap : DEFAULT_NODE_CAP;
  const [selected, setSelected] = useState(rootId);
  // The route is the source of truth for the root (back/forward, a pasted
  // URL); the selection follows it rather than pointing at a node that may
  // no longer be drawn.
  useEffect(() => { setSelected(rootId); }, [rootId]);

  const result = useMemo(() => (rootId ? walk(index, rootId, depth) : { order: [], distance: new Map(), edges: [] }), [index, rootId, depth]);
  const shown = result.order.slice(0, cap);
  const shownSet = new Set(shown);
  const nodes = shown.map((id) => index.nodes[id]);
  const edges = result.edges.filter((e) => shownSet.has(e.source) && shownSet.has(e.target));
  const truncated = result.order.length - shown.length;

  const inView = {
    "components declaring nothing": nodes.filter((n) => n.kind === "component" && n.declares === "nothing").length,
    "broken edges": edges.filter((e) => e.state === "broken").length,
    "unsatisfied edges": edges.filter((e) => e.state === "unsatisfied").length,
    "underivable edges": edges.filter((e) => e.state === "underivable").length,
    "satisfied edges": edges.filter((e) => e.state === "satisfied").length,
  };
  const c = views.state_counts;
  const corpus = {
    "components declaring nothing": `${c.components_declaring_nothing} of ${c.components}`,
    "broken edges": c.edges_broken,
    "unsatisfied edges": c.edges_unsatisfied,
    "underivable edges": c.edges_underivable,
    "satisfied edges": c.edges_satisfied,
    "components declaring in prose only": c.components_prose_only,
  };

  // Tags are the default: a reader identifying something reads a label, and
  // a label beside a dot is a scatter plot. The sphere drawing stays one
  // control away because it shows SHAPE — clusters, isolates, how far a
  // neighbourhood reaches — which a list flattens away.
  const view = route.view === "graph" ? "graph" : "tags";
  const go = (patch) => navigate({ page: "component", id: rootId, depth, cap, view, ...patch });
  const selectedNode = index.nodes[selected] || index.nodes[rootId];

  return h(React.Fragment, null,
    h("div", { className: "controls" },
      h("label", null, "component",
        h("select", { value: rootId || "", onChange: (e) => { setSelected(e.target.value); go({ id: e.target.value }); } },
          components.map((n) => h("option", { key: n.id, value: n.id }, n.label)))),
      h("label", null, "depth",
        h("input", { type: "number", min: 0, max: 8, value: depth, onChange: (e) => go({ depth: Math.max(0, Number(e.target.value) || 0) }) })),
      h("label", null, "view",
        h("select", { value: view, onChange: (e) => go({ view: e.target.value }) },
          h("option", { value: "tags" }, "diagram"),
          h("option", { value: "graph" }, "graph"))),
      h("label", null, "node cap",
        h("input", { type: "number", min: 1, max: 200, value: cap, onChange: (e) => go({ cap: Math.max(1, Number(e.target.value) || 1) }) })),
      h("span", { className: "note", "data-query-count": result.order.length, "data-shown-count": shown.length },
        `${shown.length} node${shown.length === 1 ? "" : "s"}, ${edges.length} edge${edges.length === 1 ? "" : "s"} at depth ${depth}`),
      truncated > 0 ? h("span", { className: "warn", "data-truncated": truncated },
        `showing ${shown.length} of the ${result.order.length} the query returned — nearest first; raise the node cap to see the rest`) : null,
      h("span", { className: "note" }, "click a node to select it · double-click to make it the root · drag to pan · wheel to zoom"),
      view === "tags" ? h("span", { className: "note" }, "nearest first; the badge is how many steps from the root") : null,
    ),
    h("div", { className: "page" },
      h("div", { className: "canvas" },
        h(ForceGraph, {
          nodes, edges, shape: view === "graph" ? "radial" : "flow", root: rootId, selected,
          level: (n) => result.distance.get(n.id) || 0,
          onSelect: (id) => setSelected(id),
          // Click-through: a component becomes the root of its own
          // neighbourhood. Anything else (a standard, a missing target) has
          // no neighbourhood to go to and stays selected.
          onActivate: (id) => { setSelected(id); if (index.nodes[id] && index.nodes[id].kind === "component") go({ id }); },
        })),
      h("aside", null,
        selectedNode ? h(NodePanel, { node: selectedNode, index, onNavigate: (id) => { setSelected(id); if (index.nodes[id] && index.nodes[id].kind === "component") go({ id }); } }) : null,
        h("h3", null, "in this view"), h(Counts, { rows: inView, prefix: "view" }),
        h("h3", null, "whole corpus"), h(Counts, { rows: corpus, prefix: "corpus" }),
        h(Legend, null),
      ),
    ),
  );
}

File: scripts/services/planning_ui/views/neighbourhood.py
"""The component neighbourhood — Page 2's input, and the depth-limited query
over it.

**Component level, by roll-up.** The graph's ``depends_on`` edges run mostly
phase to phase. A reader asking *what does this component depend on* wants
them attributed to components, so every edge is rolled up: a phase endpoint
becomes the component that OWNS it (the phase node's ``attrs.owner``, which
Own Phases, Referenced Phases and Edge States derived so that a status is
never read off the wrong entry), a component endpoint stays itself, and a
standard or artifact endpoint stays itself under its own kind. The phase-level
edge is kept beneath the rolled-up one as ``via``, so the panel can say WHICH
phase's line the edge was read from.

**Three states a reader must not confuse, kept apart here and drawn apart in
the browser** (requirement 4):

- a component that **declares nothing** — the accounting row's disposition,
  carried on the node as ``declares``;
- an edge that **resolves to no node** — ``state: broken``; its target is
  emitted as a node of kind ``missing`` so the edge has something to be drawn
  to, and so the reader sees a hole rather than a dropped line;
- an edge whose **target is unfinished** — ``state: unsatisfied``.

The fourth edge state, ``underivable``, is carried unchanged: the extractor
refused to guess it and this module does not guess either.

**An edge from a component to itself** — one of its phases depending on
another — does not cross the component boundary and is not a neighbourhood
edge. It is kept on the node as ``internal`` so the panel can list it and so
the corpus-wide state counts still include it.
"""
from __future__ import annotations

from collections import deque
from typing import Any

from planning_ui.plan_extractor.dependencies import (
    EXCEPTIONS,
    RESOLVES,
    WORKLIST_PROSE_ONLY,
    WORKLIST_UNDECLARED,
)
from planning_ui.plan_extractor.model import (
    EDGE_BROKEN,
    EDGE_SATISFIED,
    EDGE_UNDERIVABLE,
    EDGE_UNSATISFIED,
)

#: The order a pair of rolled-up edges collapse in: one phase edge broken and
#: another satisfied between the same two components is a broken edge between
#: them. A reader chases the worst problem first, and the `via` list keeps the
#: rest.
STATE_SEVERITY = {EDGE_BROKEN: 0, EDGE_UNSATISFIED: 1, EDGE_UNDERIVABLE: 2, EDGE_SATISFIED: 3}

#: How a node's `declares` reads, from its accounting disposition. Stated as
#: a mapping so the browser matches on five words rather than on the
#: extractor's longer disposition strings.
DECLARES = {
    RESOLVES: "resolves",
    WORKLIST_UNDECLARED: "nothing",
    WORKLIST_PROSE_ONLY: "prose-only",
}


def worst_state(states: list[str]) -> str:
    return min(states, key=lambda s: STATE_SEVERITY.get(s, 99))


def _short(path: str) -> str:
    """A component's label is its path under ``development/`` — the way the
    sprint file and every roadmap name it (``common/planning_ui``)."""
    return path[len("development/"):] if path.startswith("development/") else path


def _roll_up(node_id: str, nodes: dict[str, dict]) -> str:
    """The component-level id an edge endpoint rolls up to, or the id itself."""
    node = nodes.get(node_id)
    if node is None:
        return node_id
    if node["kind"] == "phase":
        owner = node["attrs"].get("owner")
        return f"component:{owner}" if owner else node_id
    return node_id


def component_index(graph: dict[str, Any]) -> dict[str, Any]:
    """The component-level dependency graph, keyed by node id.

    Returns ``{"nodes": {id: node}, "edges": [edge], "phases": {id: phase}}``.
    Every component in the graph is a node whether or not any edge touches it
    — a component that declares nothing has no out-edges and is still the
    thing requirement 4 asks the page to count and to draw distinctly.
    ``phases`` carries the status of every phase node, read off its owning
    entry by the extractor, so the panel can list a component's own phases
    without a second read.
    """
    nodes_by_id = {n["id"]: n for n in graph["nodes"]}
    declares_by_component = {
        f"component:{row['component']}": row for row in graph["dependency_accounting"]
    }

    index_nodes: dict[str, dict[str, Any]] = {}
    for node in graph["nodes"]:
        if node["kind"] != "component":
            continue
        row = declares_by_component.get(node["id"])
        disposition = row["disposition"] if row else ""
        if disposition in EXCEPTIONS:
            declares = "excepted"
        else:
            declares = DECLARES.get(disposition, "")
        index_nodes[node["id"]] = {
            "id": node["id"],
            "kind": "component",
            "label": _short(node["attrs"].get("path", node["label"])),
            "file": node["file"],
            "line": node["line"],
            "status": node["attrs"].get("status", ""),
            "retired": bool(node["attrs"].get("retired")),
            "declares": declares,
            "disposition": disposition,
            "owned_phases": list(node["attrs"].get("owned_phases", [])),
            "internal": [],
        }

    rolled: dict[tuple[str, str], list[dict[str, Any]]] = {}
    for edge in graph["edges"]:
        if edge["kind"] != "depends_on":
            continue
        source = _roll_up(edge["source"], nodes_by_id)
        target = _roll_up(edge["target"], nodes_by_id)
        via = {
            "source": edge["source"],
            "target": edge["target"],
            "state": edge["state"],
            "file": edge["file"],
            "line": edge["line"],
        }
        if source == target:
            if source in index_nodes:
                index_nodes[source]["internal"].append(via)
            continue
        rolled.setdefault((source, target), []).append(via)
        for endpoint in (source, target):
            if endpoint in index_nodes:
                continue
            node = nodes_by_id.get(endpoint)
            if node is None:
                # The edge's target names no node — a broken edge. It is drawn
                # to a `missing` node rather than dropped, so the reader sees
                # the hole where the target should be.
                index_nodes[endpoint] = {
                    "id": endpoint,
                    "kind": "missing",
                    "label": endpoint.split(":", 1)[-1],
                    "file": "",
                    "line": 0,
                    "status": "",
                    "retired": False,
                    "declares": "",
                    "disposition": "",
                    "owned_phases": [],
                    "internal": [],
                }
            else:
                index_nodes[endpoint] = {
                    "id": endpoint,
                    "kind": node["kind"],
                    "label": _short(node["attrs"].get("path", node["label"])),
                    "file": node["file"],
                    "line": node["line"],
                    "status": node["attrs"].get("status", ""),
                    "retired": False,
                    "declares": "",
                    "disposition": "",
                    "owned_phases": [],
                    "internal": [],
                }

    edges = [
        {
            "source": source,
            "target": target,
            "state": worst_state([v["state"] for v in vias]),
            "via": vias,
        }
        for (source, target), vias in sorted(rolled.items())
    ]
    phases = {
        n["id"]: {
            "label": n["label"],
            "file": n["file"],
            "line": n["line"],
            "status": n["attrs"].get("status", ""),
            "owner": n["attrs"].get("owner", ""),
        }
        for n in graph["nodes"]
        if n["kind"] == "phase"
    }
    return {"nodes": dict(sorted(index_nodes.items())), "edges": edges, "phases": phases}


def neighbourhood(index: dict[str, Any], component_id: str, depth: int) -> dict[str, Any]:
    """The nodes within ``depth`` edges of ``component_id``, in either
    direction, and every index edge among them.

    Breadth-first, so the returned ``order`` is nearest-first: a view that
    caps the node count keeps the nodes closest to the one the reader asked
    about. Each returned node carries its ``distance``. The browser's
    ``neighbourhood.mjs`` walks the same index the same way; this is the
    copy the tests hold to the hand-known answer.
    """
    if component_id not in index["nodes"]:
        raise KeyError(f"{component_id} is not in the neighbourhood index")
    adjacency: dict[str, set[str]] = {}
    for edge in index["edges"]:
        adjacency.setdefault(edge["source"], set()).add(edge["target"])
        adjacency.setdefault(edge["target"], set()).add(edge["source"])

    distance = {component_id: 0}
    order = [component_id]
    queue = deque([component_id])
    while queue:
        current = queue.popleft()
        if distance[current] >= depth:
            continue
        for neighbour in sorted(adjacency.get(current, ())):
            if neighbour in distance:
                continue
            distance[neighbour] = distance[current] + 1
            order.append(neighbour)
            queue.append(neighbour)

    edges = [
        e for e in index["edges"] if e["source"] in distance and e["target"] in distance
    ]
    return {"root": component_id, "depth": depth, "order": order, "distance": distance, "edges": edges}


def state_counts(graph: dict[str, Any]) -> dict[str, int]:
    """The corpus-wide counts the page states — requirement 4's three, and
    the neighbours a reader would otherwise ask about."""
    edges = [e for e in graph["edges"] if e["kind"] == "depends_on"]
    rows = graph["dependency_accounting"]
    return {
        "components": sum(1 for n in graph["nodes"] if n["kind"] == "component"),
        "components_declaring_nothing": sum(
            1 for r in rows if r["disposition"] == WORKLIST_UNDECLARED
        ),
        "components_prose_only": sum(1 for r in rows if r["disposition"] == WORKLIST_PROSE_ONLY),
        "edges_broken": sum(1 for e in edges if e["state"] == EDGE_BROKEN),
        "edges_unsatisfied": sum(1 for e in edges if e["state"] == EDGE_UNSATISFIED),
        "edges_satisfied": sum(1 for e in edges if e["state"] == EDGE_SATISFIED),
        "edges_underivable": sum(1 for e in edges if e["state"] == EDGE_UNDERIVABLE),
    }
